File: lab/may_23/solution.cc
/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <string>
#include <unordered_map>

int main ()
{
  std::string N;
  std::cin >> N;
  uint64_t m, k;
  std::cin >> m >> k;
  std::unordered_map<std::string, uint64_t> table;
  for (uint64_t i {}; i <= N.size() - k; ++i) { ++table[N.substr(i, k)]; }
  while (m--)
  {
    std::string key;
    std::cin >> key;
    if (table.find(key) != table.end()) { std::cout << table[key] << "\n"; }
    else                                { std::cout << "0\n"; }
  }
  return 0;
}

/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <map>
int main ()
{
  std::string N;
  std::cin >> N;
  uint64_t m, k;
  std::cin >> m >> k;
  std::map<uint64_t, uint64_t> table;
  for (uint64_t i {}, key {}; i != N.size(); ++i)
  {
    // this is equivalent to key = ((key * 2) + (N[i] - '0')) % 2^k
    key = ((key << 1) | (N[i] - '0')) & ((1ULL << k) - 1);
    if (i >= k - 1) { ++table[key]; }
  }
  while (m--)
  {
    std::string B;
    uint64_t key {};
    std::cin >> B;
    for (auto b : B) { key = (key << 1) | (b - '0'); }
    if (table.find(key) != table.end()) { std::cout << table[key] << "\n"; }
    else                                { std::cout << "0\n"; }
  }
  return 0;
}

File: lab/may_23/solution.py
# author: jamie
# email: jinjiedeng.jjd@gmail.com
if __name__ == "__main__":
    N = input()
    m, k = map(int, input().split())
    table = {}
    for i in range(len(N) - k + 1):
      key = N[i:i+k]
      if table.get(key) is None: table[key] = 0
      table[key] += 1
    for _ in range(m):
        B = input()
        print(0 if table.get(B) is None else table[B])

# author: jamie
# email: jinjiedeng.jjd@gmail.com
if __name__ == "__main__":
    N = input()
    m, k = map(int, input().split())
    table = {}
    key = 0
    for i, b in enumerate(N):
        # this is equivalent to key = ((key * 2) + (N[i] - '0')) % 2^k
        key = ((key << 1) | (ord(b) - ord('0'))) & ((1 << k) - 1)
        if i >= k - 1:
            if table.get(key) is None: table[key] = 0
            table[key] += 1
    for _ in range(m):
        key = 0
        B = input()
        for b in B: key = (key << 1) | (ord(b) - ord('0'))
        print(0 if table.get(key) is None else table[key])